pilot/pkg/trustbundle/trustbundle.go

```diff
@@ -0,0 +1,181 @@
+	GlobalWorkloadTrustbundle *TrustBundle = nil
+)
+
+const (
+	SourceIstioCA Source = iota
+	SourceMeshConfig
+	SourceIstioRA
+	// SOURCE_SPIFFE_ENDPOINT
+	UpdateChannelLen = 10
+)
+
+func checkSameCerts(certs1 []string, certs2 []string) bool {
+	if len(certs1) != len(certs2) {
+		return false
+	}
+	for i := range certs1 {
+		if certs1[i] != certs2[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func NewTrustBundle(updatecb func()) *TrustBundle {
+	tb := &TrustBundle{
+		sourceConfig: map[Source]TrustAnchorConfig{
+			SourceIstioCA:    {Source: SourceIstioCA, Certs: []string{}},
+			SourceMeshConfig: {Source: SourceMeshConfig, Certs: []string{}},
+			SourceIstioRA:    {Source: SourceIstioRA, Certs: []string{}},
+			// SOURCE_SPIFFE_ENDPOINT:        &TrustAnchorConfig{source: SOURCE_SPIFFE_ENDPOINT, certs: []string{}},
+		},
+		mergedCerts: []string{},
+		updatecb:    updatecb,
+	}
+
+	GlobalWorkloadTrustbundle = tb
+	return tb
+}
+
+// GetTrustBundle : Retrieves all the trustAnchors for current Spiffee Trust Domain
+func (tb *TrustBundle) GetTrustBundle() []string {
+	tb.mutex.RLock()
+	defer tb.mutex.RUnlock()
+	trustedCerts := make([]string, len(tb.mergedCerts))
+	copy(trustedCerts, tb.mergedCerts)
+	return trustedCerts
+}
+
+func (tb *TrustBundle) verifyTrustAnchor(trustAnchor string) error {
+	block, _ := pem.Decode([]byte(trustAnchor))
+	if block == nil {
+		return fmt.Errorf("failed to decode pem certificate")
+	}
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return fmt.Errorf("failed to parse X.509 certificate")
+	}
+	if !cert.IsCA {
+		return fmt.Errorf("certificate is not a CA certificate")
+	}
+	return nil
+}
+
+func (tb *TrustBundle) mergeInternal() {
+	var ok bool
+	mergeCerts := []string{}
+	certMap := make(map[string]bool)
+
+	tb.mutex.Lock()
+	defer tb.mutex.Unlock()
+
+	for _, configSource := range tb.sourceConfig {
+		for _, cert := range configSource.Certs {
+			if _, ok = certMap[cert]; !ok {
+				certMap[cert] = true
+				mergeCerts = append(mergeCerts, cert)
+			}
+		}
+	}
+	tb.mergedCerts = mergeCerts
+}
+
+// External Function to merge a TrustAnchor config with the existing TrustBundle
+// Should only be one writer
+func (tb *TrustBundle) UpdateTrustAnchor(anchorConfig *TrustAnchorUpdate) error {
+	var ok bool
+	var err error
+
+	cachedConfig, ok := tb.sourceConfig[anchorConfig.Source]
+	if !ok {
+		return fmt.Errorf("invalid source of TrustBundle configuration %v", anchorConfig.Source)
+	}
+
+	// Check if anything needs to be changed at all
+	if checkSameCerts(anchorConfig.Certs, cachedConfig.Certs) {
+		trustBundleLog.Infof("no change to trustAnchor configuration after recent update")
+		return nil
+	}
+
+	for _, cert := range anchorConfig.Certs {
+		err = tb.verifyTrustAnchor(cert)
+		if err != nil {
+			return err
+		}
+	}
+	tb.sourceConfig[anchorConfig.TrustAnchorConfig.Source] = anchorConfig.TrustAnchorConfig
+	tb.mergeInternal()
+
+	trustBundleLog.Debugf("updating Source %v with certs %v",
```

Comment:
nit: feels odd that no changes detected log is `Info`, but the updated is `Debug`. It seems like they should be swapped, since a new update is a major event

Reply:
good catch. fixed it.